core/src/heaviest_subtree_fork_choice.rs

```diff
@@ -411,6 +418,59 @@ impl HeaviestSubtreeForkChoice {
             .map(|(slot_hash, fork_info)| (slot_hash, fork_info.stake_voted_subtree))
     }
 
+    /// Dump the node `slot_hash_key` and propagate the stake subtraction up to the root of the
+    /// tree. Children of `slot_hash_key` are implicitely dumped as well, as if we were able to
+    /// chain them to a defunct parent that implies that they are defunct as well (consistent with
+    /// bank forks).
+    ///
+    /// Returns all removed slots
+    pub fn dump_node(&mut self, slot_hash_key: &SlotHashKey) -> Vec<Slot> {
+        let parent = {
+            let mut node_to_dump = self
+                .fork_infos
+                .get_mut(slot_hash_key)
+                .expect("Slot hash key must exist in tree");
+            // Remove stake to be aggregated up the tree
+            node_to_dump.stake_voted_subtree = 0;
+            node_to_dump.stake_voted_at = 0;
+            node_to_dump.is_dumped = true;
```

Comment:
is the `is_dumped` flag necessary? Seems like all nodes marked with `is_dumped = true` will just be deleted later in this function

Reply:
agreed it's a little annoying but it's used to signal for `is_candidate` so that when we aggregate to ancestors we guarantee that this node won't be chosen as the best child. We could perhaps reuse the invalid ancestor flag instead since we're most likely dumping this node because it is a duplicate anyway.

Reply:
Hmm, I'm missing where this `is_dumped` flag is being propagated to ancestors. It seems like anything marked `is_dumped  = true` here immediately gets removed below because it's added to the below `to_visit` set.

Reply:
After setting `is_dumped` we insert and process update operations (line 443). in `aggregate_slot` we have the `is_candidate` check to propagate best child which now fails because `is_dumped = true`:
```
                // Note: If there's no valid children, then the best slot should default to the
                // input `slot` itself.
                if child_fork_info.is_candidate()
                    && (best_child_slot_key == slot_hash_key ||
                    child_stake_voted_subtree > best_child_stake_voted_subtree ||
                // tiebreaker by slot height, prioritize earlier slot
                (child_stake_voted_subtree == best_child_stake_voted_subtree && child_key < &best_child_slot_key))
```

Reply:
see `test_dump_node_on_best_path` for an example that would fail without the flag, after dumping node 6 we'd expect the best child to be updated to 5.

Reply:
ah yeah, maybe reusing the existing `latest_invalid_ancestor` and just marking it  as `Some(slot_hash_key.0)` will work?

Reply:
sounds good